Agones Unity SDK development setup instructions + Agones Unity SDK Ready test

aallbrig opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/googleforgames/agones/blob/main/CONTRIBUTING.md and developer guide https://github.com/googleforgames/agones/blob/main/build/README.md
2. Please label this pull request according to what type of issue you are addressing.
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/googleforgames/agones/blob/main/build/README.md#testing-and-building
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation
/kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:
Following instructions proposed in https://github.com/googleforgames/agones/pull/3886 I set up a local environment to contribute to agones unity client sdk.

I noticed there wasn't a test. This adds a first test. I don't think it's a good test so please offer up what might make for a better first test.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Closes #<issue number>`, or `Closes (paste link of issue)`.
-->
Closes #

**Special notes for your reviewer**:

markmandel commented on sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs:
```diff
@@ -0,0 +1,33 @@
+            var sut = gameObject.AddComponent<AgonesSdk>();
+            var task = sut.Ready();
+            yield return AwaitTask(task);
+            Assert.IsTrue(task.Result);
```
This is not a bad test (although we may get double coverage if we ever do a conformance test - which is not well documented)  - but wondering if this passes, as there is no mock for the http endpoint for the `/ready` request?

igooch commented:
Noob question -- what are the steps to run the tests? Are we able to run the tests similar to vanilla csharp (csharp SDK tests have the non-obvious steps of 1. navigating to `agones/build` 2. running `make sdk-shell-csharp` 3. once in the C# shell navigating to the test folder /go/src/agones.dev/agones/sdks/csharp/test 4. using standard dotnet commands such as `dotnet test`.), or does this require a Unity editor?

ZeroParticle commented:
> Noob question -- what are the steps to run the tests? Are we able to run the tests similar to vanilla csharp (csharp SDK tests have the non-obvious steps of 1. navigating to `agones/build` 2. running `make sdk-shell-csharp` 3. once in the C# shell navigating to the test folder /go/src/agones.dev/agones/sdks/csharp/test 4. using standard dotnet commands such as `dotnet test`.), or does this require a Unity editor?

For the tests setup I added in #3883, I added a README that might help answer this for you. It's at the very bottom of the PR. My understanding is that it's basically impossible to run these tests without the unity editor. Even the common CI setups for Unity are just running a dockerized editor.

igooch commented:
> My understanding is that it's basically impossible to run these tests without the unity editor. Even the common CI setups for Unity are just running a dockerized editor.

Would it makes sense to open up a new issue for creating some sort of pipeline for running the Unity tests? Possibly CloudBuild could take in a Unity license as a Secret and run the tests in the dockerized editor mode?

aallbrig commented:
@igooch You're definitely right that developing unity packages is not intuitive. I wrote documentation in this PR (https://github.com/googleforgames/agones/pull/3886) for how to set up a testing environment that's common. I've used this setup when contributing to other unity open source projects like Mirror. You can follow the instructions in that PR to set up to then develop a unity package.

Should I merge that PR into this one, so documentation exists along with this test?

aallbrig commented:
@igooch I can help with using game CI's docker container to run tests. I've used that container image before to cross platform test & build in parallel since I didn't like serially builds in unity editor.

There's a little setup to do but I've scripted its use out before. Would need to review this project's CI more to understand how to introduce a new testing step (guidance is welcome and appreciated!)

Unity docker usage seems outside the scope of this PR, maybe, since it's a test suite agnostic enhancement? WDYT?

aallbrig commented on sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs:
```diff
@@ -0,0 +1,56 @@
+            _mockSdkServer.DeregisterResponseHandler("/ready");
+        }
+        [UnityTest]
+        public IEnumerator AgonesSdk_()
```
Done :)

aallbrig commented:
Merged in the contents of https://github.com/googleforgames/agones/pull/3886 into this PR, to provide instructions for this test.

In order to simplify the setup instructions I created a new Unity project in `test/sdk/unity`, which is very much inspired by @ZeroParticle's https://github.com/googleforgames/agones/pull/3883. The difference between this implementation and @ZeroParticle's is that I keep the test co-located with the Agones Unity SDK custom package which this new test unity project pulls in (relative path in `Packages/manifest.json`), whereas @ZeroParticle has their test in the Unity project which isn't standard to typical Unity Custom Packages (See [Unity Custom Package Layout](https://docs.unity3d.com/Manual/cus-layout.html)).

The PR is still WIP; now that I can have some guarantees with the Unity Project I can install some additional testing libraries more easily. Instead of having `sdks/unity/Tests/TestingEnvironment/MockAgonesSdkServer.cs` I could use a 3rd party library to stub out `UnityWebRequest` for tests.

aallbrig commented:
@igooch @markmandel et all This PR is ready to be reviewed!
Features:

- New Unity project created at `test/sdk/unity` that's already _pre-configured_ to pull in local Agones Unity SDK locally (see `test/sdk/unity/Package/manifest.json` so see how). This serves as the host Unity project, which is required for Unity Custom Package development.
- Instructions added on how to develop the Agones Unity SDK added so that developers know how to do things like add the `test/sdk/unity` project to their Unity Hub, and how development will work.
- A single test that spies to ensure `AgonesSdk` component makes request to the Agones SDK Server. I want to add more tests but hope for feedback. View `sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs` to view the test!


<img width="1512" alt="Screenshot 2024-07-20 at 1 42 41 PM" src="https://github.com/user-attachments/assets/a262ff7a-7a94-4296-9fcf-cd359275c7a8">

The PR is well over 1000 LoC primarily due to having a new Unity project contained in the changeset. See below to see that a majority of this PR's LoC come from default files created when initializing a new Unity project.

```
$ git diff --numstat main...agones-unity-client-sdk-initialize-test | sort -k1,1nr

671     0       test/sdk/unity/ProjectSettings/ProjectSettings.asset
322     0       test/sdk/unity/Packages/packages-lock.json
322     0       test/sdk/unity/ProjectSettings/QualitySettings.asset
316     0       test/sdk/unity/Assets/Main.unity
296     0       test/sdk/unity/ProjectSettings/InputManager.asset
121     0       test/sdk/unity/ProjectSettings/SceneTemplateSettings.json
93      0       test/sdk/unity/ProjectSettings/NavMeshAreas.asset
72      0       test/sdk/unity/.gitignore
67      0       test/sdk/unity/ProjectSettings/GraphicsSettings.asset
50      0       sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs
48      0       test/sdk/unity/ProjectSettings/Physics2DSettings.asset
47      0       test/sdk/unity/ProjectSettings/EditorSettings.asset
44      0       test/sdk/unity/Packages/manifest.json
43      0       test/sdk/unity/ProjectSettings/TagManager.asset
40      0       test/sdk/unity/ProjectSettings/DynamicsManager.asset
36      0       test/sdk/unity/ProjectSettings/PackageManagerSettings.asset
36      0       test/sdk/unity/ProjectSettings/UnityConnectSettings.asset
35      0       test/sdk/unity/ProjectSettings/MemorySettings.asset
25      0       sdks/unity/Tests/TestingEnvironment/SpyRequestSender.cs
23      0       sdks/unity/Tests/Runtime/PlayMode/Tests.Runtime.Agones.asmdef
20      0       test/sdk/unity/ProjectSettings/AudioManager.asset
18      0       test/sdk/unity/ProjectSettings/VFXManager.asset
17      12      sdks/unity/AgonesSdk.cs
14      0       sdks/README.md
14      0       sdks/unity/README.md
12      2       sdks/unity/Agones.asmdef
10      0       sdks/unity/IRequestSender.cs
9       0       test/sdk/unity/ProjectSettings/TimeManager.asset
8       0       sdks/unity/Tests/Runtime/PlayMode.meta
8       0       test/sdk/unity/ProjectSettings/EditorBuildSettings.asset
8       0       test/sdk/unity/ProjectSettings/VersionControlSettings.asset
7       0       sdks/unity/README.md.meta
7       0       test/sdk/unity/Assets/Main.unity.meta
7       0       test/sdk/unity/ProjectSettings/PresetManager.asset
6       0       test/sdk/unity/ProjectSettings/ClusterInputManager.asset
5       0       test/sdk/unity/README.md
3       0       sdks/unity/IRequestSender.cs.meta
3       0       sdks/unity/Tests.meta
3       0       sdks/unity/Tests/Runtime.meta
3       0       sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs.meta
3       0       sdks/unity/Tests/Runtime/PlayMode/Tests.Runtime.Agones.asmdef.meta
3       0       sdks/unity/Tests/TestingEnvironment.meta
3       0       sdks/unity/Tests/TestingEnvironment/SpyRequestSender.cs.meta
2       0       README.md
2       0       test/sdk/unity/ProjectSettings/ProjectVersion.txt
```
Without the `.asset`, `.meta`, `package-lock.json`, and `.unity` files this change set would probably be considered reasonable.
```
$ git diff --numstat main...agones-unity-client-sdk-initialize-test | grep -v '!*.asset\|!*.meta\|!*-lock.json\|!*\.unity' | sort -k1,1nr

121     0       test/sdk/unity/ProjectSettings/SceneTemplateSettings.json
72      0       test/sdk/unity/.gitignore
50      0       sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs
44      0       test/sdk/unity/Packages/manifest.json
25      0       sdks/unity/Tests/TestingEnvironment/SpyRequestSender.cs
23      0       sdks/unity/Tests/Runtime/PlayMode/Tests.Runtime.Agones.asmdef
17      12      sdks/unity/AgonesSdk.cs
14      0       sdks/README.md
14      0       sdks/unity/README.md
12      2       sdks/unity/Agones.asmdef
10      0       sdks/unity/IRequestSender.cs
5       0       test/sdk/unity/README.md
2       0       README.md
2       0       test/sdk/unity/ProjectSettings/ProjectVersion.txt
```

aallbrig commented:
@zmerlynn @igooch @markmandel tagging for some review. Thank you in advance!

markmandel commented on sdks/unity/IRequestSender.cs:
```diff
@@ -0,0 +1,10 @@
+using System.Threading.Tasks;
```
Apache licence header please (on everything that's code).

aallbrig replied:
Whoops! I think I need to see if I can configure my editor to automatically add these headers.

markmandel commented on sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs:
```diff
@@ -0,0 +1,50 @@
+// Copyright 2022 Google LLC
```
```suggestion
// Copyright 2024 Google LLC
```

markmandel commented on test/sdk/unity/.gitignore:
```diff
@@ -0,0 +1,72 @@
+# This .gitignore file should be placed at the root of your Unity project directory
```
Apache licence please.

markmandel commented on sdks/unity/Tests/TestingEnvironment/SpyRequestSender.cs:
```diff
@@ -0,0 +1,25 @@
+using System.Threading.Tasks;
```
Apache licence please.

markmandel commented on sdks/unity/IRequestSender.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace Agones
+{
+    public interface IRequestSender
```
oooh, what's this do? 😄 👀

markmandel commented on sdks/unity/README.md:
```diff
@@ -0,0 +1,14 @@
+# Agones Unity Client SDK
```
👍🏻  Noice.

markmandel reviewed (commented):
Thanks for doing this! Mostly nit comments.

aallbrig commented:
@markmandel I added an apache header to all added C# code in this PR. I also added it to `sdks/unity/model/StatusAddresses.cs` which is a class that pre-exists this PR but I found it was missing an apache header.

I debated if I should mark the year 2019 or 2024; I opted for 2024.

ZeroParticle commented:
Happy to update my PR #3883 to use this project if that's what we go for. The tests in my project run against a local SDK server rather than a mock. That seems complimentary to what you have here, so I figure we will want to keep it around rather than replace it.

aallbrig commented:
@ZeroParticle Thanks for weighing in! Yes I think adding tests in the style of this PR is better than in the style of your PR. The style of this PR has the tests contained within the package files which is more expected for a custom unity package.

I'm hoping @markmandel (and/or other(s)) review/merge this PR soon which will enable both of us to evolve this base code.

zmerlynn commented:
@ZeroParticle @aallbrig - @markmandel and I were chatting. We don't have a great way to test this right now, unfortunately. @ZeroParticle, if I review this, would you be up for ad hoc testing and seeing if it looks reasonable? @aallbrig - same question for other Unity PRs as they come in.

Basically, if y'all are willing, we're happy to review and then gradually let y'all be the Unity maintainers, but we may need help on the testing front.

aallbrig commented:
Was trying to get gameci docker container working but got stuck. They changed how licenses work in Unity, and a license is required to be able to use the Unity editor app. Gameci is based on Ubuntu; a hub generated license is tied to the hardware/OS -- so using my M2 license won't work for an Ubuntu container.

Unity changed the process of obtaining a license. It's generated through Unity Hub app now.

I was reverse engineering what gameci does. There's plenty of JS that obfuscates so I just need to keep reading and experimenting. There is a way of generating docker commands to run tests (I've done it before), just need to work out the license limitations.

The reason I bring up gameci is because it could help make running tests something that can run in a container, which may decrease the burden of contributing to Agones Unity SDK. Imagine recommending a user run a few commands instead of downloading/installing Unity Hub, mucking with a GUI, then opening the project in Unity Editor to then run some tests.

Just posting some updates.

I'm still hoping that this PR is deemed mergeable even without the gameci docker commands. Getting the initial Unity project set up and configured to pull in the local copy of Agones Unity SDK is foundational to further development. Even though it would be better to have shareable docker commands, we still have a way of developing/testing the Agones Unity SDK as documented in this PR.

zmerlynn commented:
I'm happy to take this as a base PR to build on. Thanks @aallbrig and @ZeroParticle !

aallbrig commented:
Woohoo!

Cc: @ZeroParticle it's been merged! Let me know if you need/want any help on your Agones Unity SDK additions/tests.
